% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/app-countOverlaps.r
\name{.countPairedBamHits}
\alias{.countPairedBamHits}
\alias{countBamHitsSingleChrom}
\alias{countNonredundant}
\alias{countPairedBamHitsSingleChrom}
\alias{countTranscriptBam}
\alias{getFeatureCounts}
\alias{getTargetRanges}
\usage{
.countPairedBamHits(input = NULL, output = NULL, param = NULL)

countPairedBamHitsSingleChrom(chr, bamFile = NULL, param = NULL,
  gff = NULL)

countNonredundant(bamFile, param = param, gff = gff)

countBamHitsSingleChrom(chr, bamFile = NULL, param = NULL, gff = NULL)

getTargetRanges(gff, param, chrom = NULL)

getFeatureCounts(chrom, gff, reads, param)

countTranscriptBam(bamFile, isFirstMateRead = NA, isSecondMateRead = NA,
  isUnmappedQuery = FALSE, isProperPair = NA)
}
\section{Functions}{
\itemize{
\item \code{.countPairedBamHits}: Counts the paired BAM hits.

\item \code{countPairedBamHitsSingleChrom}: Counts the paired BAM hits for a single chromosome.

\item \code{countNonredundant}: Counts the non-redundant overlaps.

\item \code{countBamHitsSingleChrom}: Counts the BAM hits for a single chromosome.

\item \code{getTargetRanges}: Gets the target ranges depending on \code{param$featureLevel}.

\item \code{getFeatureCounts}: Gets the feature counts from the target ranges.

\item \code{countTranscriptBam}: Counts the transcripts from a BAM file.
}}

